File: packages/ld-adapter/src/__tests__/apply-targeting.test.ts
import { readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { afterEach, describe, expect, it, vi } from 'vitest';
import { applyTargeting } from '../write/apply-targeting.js';
import { SEMANTIC_PATCH_CONTENT_TYPE } from '../client/ld-api-client.js';
import { UnresolvedVariationError } from '../errors/ld-adapter-error.js';
import type { LaunchDarklyRawClient } from '../client/ld-api-client.js';
import LaunchDarklyApi from 'launchdarkly-api';

const fixturesDir = join(dirname(fileURLToPath(import.meta.url)), 'fixtures');
const flagFixture = JSON.parse(
	readFileSync(join(fixturesDir, 'flag-boolean.json'), 'utf8'),
);

function createMockStack() {
	const getFeatureFlag = vi.fn((_project, _flag, _opts, callback) => {
		callback(null, flagFixture);
	});
	const fetchMock = vi.fn().mockResolvedValue({
		ok: true,
		status: 200,
		json: async () => flagFixture,
		headers: { entries: () => [][Symbol.iterator]() },
	});

	const rawClient = {
		flagsApi: {
			getFeatureFlag,
			apiClient: LaunchDarklyApi.ApiClient.instance,
		},
		config: {
			accessToken: 'test-token',
			baseUrl: 'https://app.launchdarkly.com',
			apiVersion: '20240415',
		},
	} as unknown as LaunchDarklyRawClient;

	const scheduleSpy = vi.fn(async <T>(fn: () => Promise<T>) => fn());
	const rateLimitedClient = {
		schedule: scheduleSpy,
		rawClient,
	};

	vi.stubGlobal('fetch', fetchMock);

	return { rawClient, rateLimitedClient, getFeatureFlag, fetchMock, scheduleSpy };
}

describe('applyTargeting', () => {
	afterEach(() => {
		vi.unstubAllGlobals();
	});

	it('GET-before-write: getFeatureFlag before fetch PATCH', async () => {
		const { rateLimitedClient, getFeatureFlag, fetchMock } = createMockStack();
		const callOrder: string[] = [];
		getFeatureFlag.mockImplementation((_p, _f, _o, cb) => {
			callOrder.push('get');
			cb(null, flagFixture);
		});
		fetchMock.mockImplementation(async () => {
			callOrder.push('patch');
			return {
				ok: true,
				status: 200,
				json: async () => flagFixture,
				headers: { entries: () => [][Symbol.iterator]() },
			};
		});

		await applyTargeting(
			{ rateLimitedClient },
			{
				projectKey: 'default',
				flagKey: 'sample-feature',
				intent: {
					environmentKey: 'production',
					rollout: {
						mode: 'fallthrough',
						treatmentVariationRef: { by: 'value', value: true },
						controlVariationRef: { by: 'value', value: false },
						treatmentPercentThousandths: 10_000,
						rolloutContextKind: 'user',
						rolloutBucketBy: 'user',
					},
				},
			},
		);

		expect(callOrder[0]).toBe('get');
		expect(callOrder).toContain('patch');
		expect(getFeatureFlag.mock.invocationCallOrder[0]).toBeLessThan(
			fetchMock.mock.invocationCallOrder[0],
		);
	});

	it('routes GET and PATCH through rateLimitedClient.schedule', async () => {
		const { rateLimitedClient, scheduleSpy } = createMockStack();
		await applyTargeting(
			{ rateLimitedClient },
			{
				projectKey: 'default',
				flagKey: 'sample-feature',
				intent: { environmentKey: 'production', turnOn: true },
			},
		);
		expect(scheduleSpy.mock.calls.length).toBeGreaterThanOrEqual(2);
	});

	it('sets semantic patch Content-Type on fetch PATCH', async () => {
		const { rateLimitedClient, fetchMock } = createMockStack();
		await applyTargeting(
			{ rateLimitedClient },
			{
				projectKey: 'default',
				flagKey: 'sample-feature',
				intent: { environmentKey: 'production', turnOn: true },
			},
		);
		const init = fetchMock.mock.calls[0]?.[1] as RequestInit;
		expect(init.headers).toMatchObject({
			'Content-Type': SEMANTIC_PATCH_CONTENT_TYPE,
		});
	});

	it('throws UnresolvedVariationError before patch when ref cannot resolve', async () => {
		const { rateLimitedClient, fetchMock } = createMockStack();
		await expect(
			applyTargeting(
				{ rateLimitedClient },
				{
					projectKey: 'default',
					flagKey: 'sample-feature',
					intent: {
						environmentKey: 'production',
						rollout: {
							mode: 'fallthrough',
							treatmentVariationRef: { by: 'value', value: 'missing' },
							controlVariationRef: { by: 'value', value: false },
							treatmentPercentThousandths: 10_000,
							rolloutContextKind: 'user',
							rolloutBucketBy: 'user',
						},
					},
				},
			),
		).rejects.toBeInstanceOf(UnresolvedVariationError);
		expect(fetchMock).not.toHaveBeenCalled();
	});

	it('uses resolveRuleId when rollout.mode is rule', async () => {
		const { rateLimitedClient, fetchMock } = createMockStack();
		await applyTargeting(
			{ rateLimitedClient },
			{
				projectKey: 'default',
				flagKey: 'sample-feature',
				intent: {
					environmentKey: 'production',
					rollout: {
						mode: 'rule',
						ruleRef: { by: 'id', id: 'rule-prod-1' },
						treatmentVariationRef: { by: 'value', value: true },
						controlVariationRef: { by: 'value', value: false },
						treatmentPercentThousandths: 10_000,
						rolloutContextKind: 'user',
						rolloutBucketBy: 'user',
					},
				},
			},
		);

		const init = fetchMock.mock.calls[0]?.[1] as RequestInit;
		const patchBody = JSON.parse(String(init.body)) as {
			instructions: Array<{ kind: string; ruleId?: string }>;
		};
		const ruleInstruction = patchBody.instructions.find(
			(i) => i.kind === 'updateRuleVariationOrRollout',
		);
		expect(ruleInstruction?.ruleId).toBe('rule-prod-1');
	});
});

File: kotlin/modules/ld-adapter/src/test/kotlin/com/ffpromo/ldadapter/write/ApplyTargetingTest.kt
package com.ffpromo.ldadapter.write

import com.ffpromo.contracts.ApplyTargetingInput
import com.ffpromo.contracts.RolloutIntent
import com.ffpromo.contracts.TargetingIntent
import com.ffpromo.contracts.VariationRef
import com.ffpromo.ldadapter.client.LaunchDarklyRawClient
import com.ffpromo.ldadapter.client.RateLimitedLdClient
import com.ffpromo.ldadapter.client.ResolvedLaunchDarklyConfig
import com.ffpromo.ldadapter.client.SEMANTIC_PATCH_CONTENT_TYPE
import com.ffpromo.ldadapter.errors.UnresolvedVariationError
import com.launchdarkly.api.ApiClient
import com.launchdarkly.api.api.FeatureFlagsApi
import kotlinx.coroutines.test.runTest
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonPrimitive
import okhttp3.OkHttpClient
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows

class ApplyTargetingTest {
    private lateinit var server: MockWebServer
    private val json = Json { ignoreUnknownKeys = true }

    @BeforeEach
    fun setUp() {
        server = MockWebServer()
        server.start()
    }

    @AfterEach
    fun tearDown() {
        server.shutdown()
    }

    private fun fixture(name: String): String =
        javaClass.getResource("/fixtures/$name")!!.readText()

    private fun createStack(): TestStack {
        val flagFixture = fixture("flag-boolean.json")
        val rawClient = LaunchDarklyRawClient(
            flagsApi = FeatureFlagsApi(ApiClient()),
            config = ResolvedLaunchDarklyConfig(
                accessToken = "test-token",
                baseUrl = server.url("/").toString().trimEnd('/'),
                apiVersion = "20240415",
            ),
            httpClient = OkHttpClient(),
        )
        val rateLimitedClient = RateLimitedLdClient.create(rawClient)
        return TestStack(rawClient, rateLimitedClient, flagFixture)
    }

    private data class TestStack(
        val rawClient: LaunchDarklyRawClient,
        val rateLimitedClient: RateLimitedLdClient,
        val flagFixture: String,
    )

    @Test
    fun `GET-before-write getFeatureFlag before PATCH`() = runTest {
        val stack = createStack()
        val callOrder = mutableListOf<String>()

        server.enqueue(MockResponse().setResponseCode(200).setBody(stack.flagFixture))
        server.enqueue(MockResponse().setResponseCode(200).setBody("{}"))
        server.enqueue(MockResponse().setResponseCode(200).setBody(stack.flagFixture))

        applyTargeting(
            ApplyTargetingDeps(stack.rateLimitedClient),
            ApplyTargetingInput(
                projectKey = "default",
                flagKey = "sample-feature",
                intent = TargetingIntent(
                    environmentKey = "production",
                    rollout = RolloutIntent(
                        mode = "fallthrough",
                        treatmentVariationRef = VariationRef.ByValue(JsonPrimitive(true)),
                        controlVariationRef = VariationRef.ByValue(JsonPrimitive(false)),
                        treatmentPercentThousandths = 10_000,
                        rolloutContextKind = "user",
                        rolloutBucketBy = "user",
                    ),
                ),
            ),
        )

        assertEquals("GET", server.takeRequest().method)
        assertEquals("PATCH", server.takeRequest().method)
        assertEquals("GET", server.takeRequest().method)
    }

    @Test
    fun `routes GET and PATCH through rateLimitedClient schedule`() = runTest {
        val stack = createStack()
        server.enqueue(MockResponse().setResponseCode(200).setBody(stack.flagFixture))
        server.enqueue(MockResponse().setResponseCode(200).setBody("{}"))
        server.enqueue(MockResponse().setResponseCode(200).setBody(stack.flagFixture))

        applyTargeting(
            ApplyTargetingDeps(stack.rateLimitedClient),
            ApplyTargetingInput(
                projectKey = "default",
                flagKey = "sample-feature",
                intent = TargetingIntent(environmentKey = "production", turnOn = true),
            ),
        )

        assertEquals(3, server.requestCount)
    }

    @Test
    fun `sets semantic patch Content-Type on PATCH`() = runTest {
        val stack = createStack()
        server.enqueue(MockResponse().setResponseCode(200).setBody(stack.flagFixture))
        server.enqueue(MockResponse().setResponseCode(200).setBody("{}"))
        server.enqueue(MockResponse().setResponseCode(200).setBody(stack.flagFixture))

        applyTargeting(
            ApplyTargetingDeps(stack.rateLimitedClient),
            ApplyTargetingInput(
                projectKey = "default",
                flagKey = "sample-feature",
                intent = TargetingIntent(environmentKey = "production", turnOn = true),
            ),
        )

        server.takeRequest()
        val patchRequest = server.takeRequest()
        assertEquals("PATCH", patchRequest.method)
        assertTrue(patchRequest.getHeader("Content-Type")!!.startsWith(SEMANTIC_PATCH_CONTENT_TYPE))
    }

    @Test
    fun `throws UnresolvedVariationError before patch when ref cannot resolve`() = runTest {
        val stack = createStack()
        server.enqueue(MockResponse().setResponseCode(200).setBody(stack.flagFixture))

        assertThrows<UnresolvedVariationError> {
            applyTargeting(
                ApplyTargetingDeps(stack.rateLimitedClient),
                ApplyTargetingInput(
                    projectKey = "default",
                    flagKey = "sample-feature",
                    intent = TargetingIntent(
                        environmentKey = "production",
                        rollout = RolloutIntent(
                            mode = "fallthrough",
                            treatmentVariationRef = VariationRef.ByValue(JsonPrimitive("missing")),
                            controlVariationRef = VariationRef.ByValue(JsonPrimitive(false)),
                            treatmentPercentThousandths = 10_000,
                            rolloutContextKind = "user",
                            rolloutBucketBy = "user",
                        ),
                    ),
                ),
            )
        }

        assertEquals(1, server.requestCount)
    }

    @Test
    fun `uses resolveRuleId when rollout mode is rule`() = runTest {
        val stack = createStack()
        server.enqueue(MockResponse().setResponseCode(200).setBody(stack.flagFixture))
        server.enqueue(MockResponse().setResponseCode(200).setBody("{}"))
        server.enqueue(MockResponse().setResponseCode(200).setBody(stack.flagFixture))

        applyTargeting(
            ApplyTargetingDeps(stack.rateLimitedClient),
            ApplyTargetingInput(
                projectKey = "default",
                flagKey = "sample-feature",
                intent = TargetingIntent(
                    environmentKey = "production",
                    rollout = RolloutIntent(
                        mode = "rule",
                        ruleRef = com.ffpromo.contracts.RuleRef.ById("rule-prod-1"),
                        treatmentVariationRef = VariationRef.ByValue(JsonPrimitive(true)),
                        controlVariationRef = VariationRef.ByValue(JsonPrimitive(false)),
                        treatmentPercentThousandths = 10_000,
                        rolloutContextKind = "user",
                        rolloutBucketBy = "user",
                    ),
                ),
            ),
        )

        server.takeRequest()
        val patchBody = server.takeRequest().body.readUtf8()
        assertTrue(patchBody.contains("updateRuleVariationOrRollout"))
        assertTrue(patchBody.contains("rule-prod-1"))
    }
}
